src/misc/messages.cpp

```diff
@@ -62,56 +84,79 @@ class Message {
 		if (rendered_msg.empty())
 			rendered_msg = convert_ansi_markup(markup_msg.c_str());
 		assert(rendered_msg.length());
-		return rendered_msg.c_str();
+
+		if (!is_utf8_msg)
+			return rendered_msg.c_str();
+
+		const uint16_t cp = UTF8_GetCodePage();
+		if (rendered_cp_msg[cp].empty()) {
+			if (!UTF8_RenderForDos(rendered_msg, rendered_cp_msg[cp], cp))
+				LOG_WARNING("LANG: Problem rendering string");
+			assert(rendered_cp_msg[cp].length());
+		}
```

Comment:
Code duplication; could extract this to a lambda with a few parameters.

Reply:
Extracted to a separate private method.